# Episource- Assignment
### By Himanshu Daga
Assignment-

    Reference File: http://textfiles.com/stories/3gables.txt
    Targets:
    1. Generate counts of words from the file above
    2. Print a list of Top 5 and Bottom 5 words
    3. Create a pandas data frame with words and their counts as two columns
    4. Remove words with count less than 5
    5. Create a new column with conditional - if the length of the word is more than 5 and count more than 10

### 1.1 Reading the text-file

In [291]:
# Get the text file as a string.
import urllib.request, urllib.parse, urllib.error
import pandas as pd
import re
import string

url='http://textfiles.com/stories/3gables.txt'
fhand = urllib.request.urlopen(url)

data=''
for line in fhand:
    data += line.decode().strip()+ ' '
# Convert the file to all lower case
data=data.lower() 
data

':.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.: -----=====earth\'s dreamlands=====----- (313)558-5024 {14.4} (313)558-5517 a bbs for text file junkies rpgnet gm file archive site .:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.  the adveniure of the three gables  i don\'t think that any of my adventures with mr. sherlock holmes opened quite so abruptly, or so dramatically, as that which i associate with the three gables. i had not seen holmes for some days and had no idea of the new channel into which his activities had been directed. he was in a chatty mood that morning, however, and had just settled me into the well-worn low armchair on one side of the fire, while he had curled down with his pipe in his mouth upon the opposite chair, when our visitor arrived. if i had said that a mad bull had arrived it would give a clearer impression of what occurred. the door had flown open and a huge negro had burst into the room. he would have been a comic figure if he

In [258]:
#remove the end of file character
data=data[:-2]
data

':.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.: -----=====earth\'s dreamlands=====----- (313)558-5024 {14.4} (313)558-5517 a bbs for text file junkies rpgnet gm file archive site .:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.:.  the adveniure of the three gables  i don\'t think that any of my adventures with mr. sherlock holmes opened quite so abruptly, or so dramatically, as that which i associate with the three gables. i had not seen holmes for some days and had no idea of the new channel into which his activities had been directed. he was in a chatty mood that morning, however, and had just settled me into the well-worn low armchair on one side of the fire, while he had curled down with his pipe in his mouth upon the opposite chair, when our visitor arrived. if i had said that a mad bull had arrived it would give a clearer impression of what occurred. the door had flown open and a huge negro had burst into the room. he would have been a comic figure if he

### 1.2 Pre-processing
    -Remove all things except alphabets

In [273]:
words= re.findall("[a-z]+", data)
counts={}
for w in words:
    counts[w]= counts.get(w,0) + 1
d=list(counts.items())
df=pd.DataFrame(d, columns=['word','count'])

In [260]:
# first 5 words
df.loc[:5] #df.head(5)

word  count
0       earth      2
1           s     41
2  dreamlands      1
3           a    147
4         bbs      1
5         for     49

In [261]:
# last 5 words
df.iloc[-5:] #df.tail(5)

word  count
1428    edged      1
1429    tools      1
1430  forever      1
1431  cutting      1
1432   dainty      1

 Above code splitted words like "Earth's" into 2 words- "Earth" & "s".
 
 To avoid such things, let's do some step-by-step preprocessing where we remove all the numbers and all the punctuations.

In [289]:
# remove numbers
text= re.sub(r'\d+', '', data)
# remove punctuations
text=text.translate(text.maketrans("","", string.punctuation))
#total no. of words
len(text.split())

6096

### 1.3 Count the words

In [277]:
# let's count all these words
counts={}
for w in text.split():
    counts[w]= counts.get(w,0) + 1
d=list(counts.items())

## 2. Print top 5 and bottom 5 words

In [278]:
# get top 5 and bottom 5 words
print("Top 5: " + str(d[:5]))
print("Bottom 5: " + str(d[-5:]))

# get 5 most frequent words and 5 least frequent words
a=sorted( [ (v,k) for k,v in counts.items() ], reverse= True )
print("5 most frequent: " + str(a[:5]))
print("5 least frequent: " + str(a[-5:]))

Top 5: [('earths', 1), ('dreamlands', 1), ('a', 145), ('bbs', 1), ('for', 49)]
Bottom 5: [('edged', 1), ('tools', 1), ('forever', 1), ('cutting', 1), ('dainty', 1)]
5 most frequent: [(264, 'the'), (198, 'i'), (145, 'a'), (143, 'you'), (141, 'to')]
5 least frequent: [(1, 'across'), (1, 'acre'), (1, 'accounts'), (1, 'abruptly'), (1, 'above')]


## 3. Create dataframe

In [283]:
# create dataframe
df=pd.DataFrame(d, columns=['word','frequency'])
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1451 entries, 0 to 1450
Data columns (total 2 columns):
word         1451 non-null object
frequency    1451 non-null int64
dtypes: int64(1), object(1)
memory usage: 22.8+ KB
None


word  frequency
0      earths          1
1  dreamlands          1
2           a        145
3         bbs          1
4         for         49

## 4. Remove words with count less than 5

In [284]:
df1= df[df['frequency']>4]
print(df1.info())
df1.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 204 entries, 2 to 1070
Data columns (total 2 columns):
word         204 non-null object
frequency    204 non-null int64
dtypes: int64(1), object(1)
memory usage: 4.8+ KB
None


word  frequency
2       a        145
4     for         49
12    the        264
14     of        128
15  three          9

In [266]:
# reset indices
df1= df1.reset_index()
df1= df1.drop('index',axis=1)
df1.head()

word  frequency
0      a        145
1    for         49
2    the        264
3     of        128
4  three          9

## 5. Apply given condition
Condition:- if word length>5 and word frequency>10

In [271]:
df1['condition']=False
for i in range(df1.shape[0]):
    if len(df1.word[i])>5 and df1.frequency[i]>10:
        df1.loc[i,'condition']=True
df1.head()

word  frequency  condition
0      a        145      False
1    for         49      False
2    the        264      False
3     of        128      False
4  three          9      False

In [272]:
# All words satisfying the given condition
df1[df1.condition]

word  frequency  condition
14     holmes         70       True
60     masser         17       True
134    matter         12       True
151  maberley         12       True
181   nothing         11       True

## Summary:
1. Total number of words in the document- 6096
2. Total number of distinct words- 1451
3. Total number of words with frequency >= 5 is- 204
4. Total number of words with frequency > 10 and length > 5 is- 5

Improvement scope:
We can do:
1. ***Stop words removal*** on our text corpus to remove less important words like 'a', 'the', 'on' etc.
2. ***Stemming and Lemmatization*** to treat words like look & looked same and different forms of a word like be & been same for the purpose of counting
For this we can use NLTK library.

##   Thank You !